source/extensions/quic_listeners/quiche/envoy_quic_stream.h

```diff
@@ -1,41 +1,95 @@
 #pragma once
 
+#include "envoy/event/dispatcher.h"
 #include "envoy/http/codec.h"
 
 #include "common/http/codec_helper.h"
 
+#include "extensions/quic_listeners/quiche/envoy_quic_simulated_watermark_buffer.h"
+
 namespace Envoy {
 namespace Quic {
 
 // Base class for EnvoyQuicServer|ClientStream.
 class EnvoyQuicStream : public Http::StreamEncoder,
                         public Http::Stream,
-                        public Http::StreamCallbackHelper {
+                        public Http::StreamCallbackHelper,
+                        protected Logger::Loggable<Logger::Id::quic_stream> {
 public:
+  EnvoyQuicStream(uint32_t buffer_limit, std::function<void()> below_low_watermark,
+                  std::function<void()> above_high_watermark)
+      : send_buffer_simulation_(buffer_limit / 2, buffer_limit, std::move(below_low_watermark),
+                                std::move(above_high_watermark), ENVOY_LOGGER()) {}
+
   // Http::StreamEncoder
   Stream& getStream() override { return *this; }
 
   // Http::Stream
+  void readDisable(bool disable) override {
+    bool status_changed{false};
+    if (disable) {
+      ++read_disable_counter_;
+      ASSERT(read_disable_counter_ == 1);
```

Comment:
The docs (and Envoy flow control) allows readDisable(true) to be called multiple times.  I'd think the second time, read_disable_counter would be 2.